src/shared/sol-conffile.c

```diff
@@ -58,147 +61,187 @@ sol_conffile_entry_sort_cb(const void *data1, const void *data2)
-        SOL_DBG("could not read value from %s group: %s",
-            include_group, error->message);
-        g_error_free(error);
+    SOL_JSON_SCANNER_OBJECT_LOOP (&json_scanner, &token, &key, &value, reason) {
+        if (sol_json_token_str_eq(&key, include_group, strlen(include_group))) {
+            found_include = true;
+            break;
+        }
     }
 
-    if (!g_key_file_remove_group(keyfile, include_group, &error)) {
-        SOL_DBG("could not remove %s group: %s", include_group,
-            error ? error->message : "unknown error");
-        g_error_free(error);
+    if (!found_include)
+        return;
+
+    sol_json_scanner_init_from_token(&include_scanner, &value);
+    SOL_JSON_SCANNER_OBJECT_LOOP (&include_scanner, &token, &key, &value, reason) {
+        if (sol_json_token_str_eq(&key, include_str, strlen(include_str))) {
+            *include = strndup(value.start, sol_json_token_get_size(&value));
+            if (!*include)
+                SOL_DBG("Error: couldn't allocate memory for string.");
+            continue;
+        }
+        if (sol_json_token_str_eq(&key, include_fallback, strlen(include_fallback))) {
+            *include_fallbacks = strndup(value.start, sol_json_token_get_size(&value));
+            if (!*include_fallbacks)
+                SOL_DBG("Error: couldn't allocate memory for string.");
+            continue;
+        }
     }
 }
 
-static GKeyFile *
+static char *
 sol_conffile_load_keyfile_from_dirs(const char *file, char **full_path)
 {
-    char *cwd = NULL;
-    GError *error = NULL;
-    GKeyFile *keyfile = NULL;
-    char cwdbuf[PATH_MAX + 1];
+    int i;
+    char *config_file_contents;
     const char *search_dirs[] = {
         NULL,
         ".",
         PKGSYSCONFDIR,
         NULL,
     };
+    int search_dirs_size;
 
-    cwd = getcwd(cwdbuf, sizeof(cwdbuf));
-    if (cwd)
-        search_dirs[0] = cwd;
-    else
-        SOL_DBG("could not getcwd: %d", errno);
-
-    keyfile = g_key_file_new();
-    SOL_NULL_CHECK(keyfile, NULL);
+    search_dirs_size = ARRAY_SIZE(search_dirs);
 
     /* absolute path */
-    if (g_path_is_absolute(file)) {
-        if (!g_key_file_load_from_file(keyfile,
-            file, G_KEY_FILE_NONE, &error))
-            goto err;
-    } else if (!g_key_file_load_from_dirs(keyfile, file, search_dirs,
-        full_path, G_KEY_FILE_NONE,
-        &error))
-        goto err;
-
-    return keyfile;
-
-err:
-    SOL_DBG("could not load file '%s' from dirs: %s",
-        file, error ? error->message : "unknown error");
-    g_error_free(error);
-    g_key_file_free(keyfile);
-    return NULL;
+    if (file[0] == '/') {
+        config_file_contents = sol_util_load_file_string(file, NULL);
+    } else {
+        for (i = 0; i < search_dirs_size; i++) {
+            char filename[PATH_MAX];
+            size_t size;
+            sprintf(filename, "%s/%s", search_dirs[i], file);
+            config_file_contents = sol_util_load_file_string(file, &size);
```

Comment:
Since you are dealing with regular files for sure, Instead of use `sol_util_load_file_string` I think you should use `sol_file_reader_open` and sol_file_reader API to deal with files that you will parse. `sol_file_reader` API will take advantage of mmap whenever its possible (i.e. when dealing with regular files which is the case) and fallback to `sol_util_load_file_raw` only if mmap isn't possible. Than just use sol_file_reader_get_all and you will have proper slice to feed `sol_json_scanner_init` (slice.data). You can check `_json_open_doc` in 'sol-platform-detect.c' if needed.

Since our parser use silces everywhere, you dont need to `sol_util_load_file_string` anyway since you dont need to enforce the contents to be null terminated (only reason to use _string instead of _raw).